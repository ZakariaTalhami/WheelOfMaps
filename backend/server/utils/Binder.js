/**
 * Binds all the methods within a class to "this"
 */
class Binder {}

/**
 * Get all the methods in the given instance
 * @param {Object} instance
 * @return {Array} - list of methods
 */
Binder.getAllMethods = function (instance) {
    var props = [];
    var obj = instance;
    while ((obj = Object.getPrototypeOf(obj)) && obj != Object.prototype) {
        props = props.concat(Object.getOwnPropertyNames(obj));
    }

    return props.sort().filter(function (e, i, arr) {
        if (e != arr[i + 1] && typeof instance[e] == "function") return true;
    });
};

/**
 * Binds the instance's methods
 * @param {Object} instance
 */
Binder.bind = function (instance) {
    Binder.getAllMethods(instance).forEach((mtd) => {
        instance[mtd] = instance[mtd].bind(instance);
    });
};

export default Binder;
