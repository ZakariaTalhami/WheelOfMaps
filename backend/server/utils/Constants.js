export const CHAPTER_IDNEX_RANGE_REGEX = /^\d{5}(-\d{5})?$/;
export const CHAPTER_RANGE_SCHEMA = {
    type: String,
    required: true,
    match: CHAPTER_IDNEX_RANGE_REGEX,
};
