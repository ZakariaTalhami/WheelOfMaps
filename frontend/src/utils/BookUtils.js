/*
    Uility module for common functionality for Books and chapters.
*/

export const getBookChapterTitles = (book) => {
    const chapters = book.chapters || [];
    return chapters.map((chapter) => {
        const prefix =
            chapter.number === 0 ? "Prologue" : `Chapter ${chapter.number}`;

        return `${prefix}: ${chapter.title}`;
    });
};
