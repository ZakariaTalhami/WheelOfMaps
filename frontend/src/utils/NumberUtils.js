/**
 * Pads a number depending on the number of digits required
 * Returns a string with the number padded with zeros
 *
 * @param {Number} num - the desired number
 * @param {Number} places - number of digits
 * @returns {String} formated string with zero padding
 */
export function zeroPad(num, places) {
    var zero = places - num.toString().length + 1;
    return Array(+(zero > 0 && zero)).join("0") + num;
}
