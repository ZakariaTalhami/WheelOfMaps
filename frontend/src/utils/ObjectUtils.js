import _ from "lodash";

const evaluateObjectPath = (obj, path = [], defaultValue) => {
    if (_.isString(path)) path = _.split(path, ".");
    if (!_.isObject(obj) || _.isEmpty(path)) return obj || defaultValue;

    return evaluateObjectPath(obj[_.head(path)], _.drop(path), defaultValue);
};

const ObjectUtils = {
    evaluateObjectPath,
};

export default ObjectUtils;
