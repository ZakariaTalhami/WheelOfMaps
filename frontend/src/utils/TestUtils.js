import React from "react";
import { render } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
//Global Theme
import extendedTheme from "../assets/styles/extendedTheme";
// TODO: remove custom font icons, replace with react-icons
import "../assets/fonts/style.css";
// Leafletjs Styles
import "leaflet/dist/leaflet.css";
import "../assets/_leafletjs.override.scss";

const AllTheProviders = ({ children }) => {
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    return <ChakraProvider theme={extendedTheme}>{children}</ChakraProvider>;
};

const customRender = (ui, options) =>
    render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
