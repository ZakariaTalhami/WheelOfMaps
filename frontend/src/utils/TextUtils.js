import { Text } from "@chakra-ui/react";

/**
 * Split the summary by new line and wrap in paragraphs
 * @param {String} summary
 * @returns {Array} paragraphs of the summary
 */
export const getSummaryParagraphs = (summary) => {
    const paragraphs = summary.split("\r\n\r\n");
    return paragraphs.map((para, index) => (
        <Text
            w="100%"
            role="paragraph"
            style={{ textIndent: "1.5rem" }}
            key={index}
        >
            {para}
        </Text>
    ));
};
