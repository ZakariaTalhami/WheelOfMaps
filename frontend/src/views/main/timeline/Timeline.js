// React and utils
import React from "react";

// Redux Actions
import {
    setSelectedBook,
    setSelectedChapter,
} from "../../../actions/BookActions";

// Components
import BookCrumbs from "./bookCrumbs/BookCrumbs";
import NavButton from "./navButton/NavButton";
import { Flex } from "@chakra-ui/react";

// Hooks
import { useDispatch, useSelector } from "react-redux";

const Timeline = () => {
    const bookState = useSelector((state) => state.Books);
    const dispatch = useDispatch();

    const handleBack = () => {
        if (bookState.selectedChapter - 1 > -1) {
            // Go to the previous chapter
            dispatch(setSelectedChapter(bookState.selectedChapter - 1));
        } else {
            // Go to the previous book
            const bookTitles = Object.keys(bookState.books);
            const bookIndex = bookTitles.indexOf(bookState.selectedBook);
            if (bookIndex - 1 > -1) {
                dispatch(setSelectedBook(bookTitles[bookIndex - 1]));
            }
        }
    };

    const handleNext = () => {
        if (
            bookState.selectedChapter + 1 <
            bookState.books[bookState.selectedBook].chapters.length
        ) {
            // Go to the next chapter
            dispatch(setSelectedChapter(bookState.selectedChapter + 1));
        } else {
            // Go to the next book
            const bookTitles = Object.keys(bookState.books);
            const bookIndex = bookTitles.indexOf(bookState.selectedBook);
            if (bookIndex + 1 < bookTitles.length) {
                dispatch(setSelectedBook(bookTitles[bookIndex + 1]));
            }
        }
    };

    return (
        <Flex
            bgColor="neutralColor"
            pointerEvents="all"
            boxShadow="inset -30px 0px 20px rgba(0, 0, 0, 0.25)"
        >
            {/* Back button */}
            <NavButton
                data-testid="back-button"
                aria-label="Previous Chapter"
                onClick={handleBack}
                icon={<i className="icon-back"></i>}
            />
            {/* Book and Chapter Labels / dropdowns */}
            <BookCrumbs {...bookState} />
            {/* Next button */}
            <NavButton
                data-testid="next-button"
                aria-label="Next Chapter"
                onClick={handleNext}
                icon={<i className="icon-next"></i>}
            />
        </Flex>
    );
};

export default Timeline;
