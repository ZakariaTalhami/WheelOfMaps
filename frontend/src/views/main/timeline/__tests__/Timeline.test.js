import React from "react";
import {
    setSelectedBook,
    setSelectedChapter,
} from "../../../../actions/BookActions";
import { fireEvent, render, screen } from "../../../../utils/TestUtils";
import Timeline from "../Timeline";
import { useSelector } from "react-redux";

const MOCK_BOOKS = {
    "Book 1": {
        chapters: [
            {
                number: 1,
                title: "title 1",
            },
            {
                number: 2,
                title: "title 1",
            },
        ],
    },
    "Book 2": {
        chapters: [
            {
                number: 1,
                title: "title 2",
            },
            {
                number: 2,
                title: "title 2",
            },
        ],
    },
};

const MOCK_BOOK_STATE = {
    books: MOCK_BOOKS,
    selectedBook: "Book 1",
    selectedChapter: 0,
};

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
    useSelector: jest.fn(),
    useDispatch: () => mockDispatch,
}));

beforeEach(() => {
    useSelector.mockReturnValue(MOCK_BOOK_STATE);
});

test("renders", () => {
    render(<Timeline />);

    // navigation buttons render
    expect(screen.getAllByRole("button").length).toEqual(2);
    expect(screen.getByLabelText("Next Chapter")).toBeDefined();
    expect(screen.getByLabelText("Previous Chapter")).toBeDefined();

    // selects render
    expect(screen.getByTestId("book-select")).toBeDefined();
    expect(screen.getByTestId("chapter-select")).toBeDefined();
});

test("handle back action - stop at first book", () => {
    render(<Timeline />);

    fireEvent.click(screen.getByTestId("back-button"));

    expect(mockDispatch).not.toHaveBeenCalled();
});

test("handle back action - goes to previous book", () => {
    useSelector.mockReturnValue({
        ...MOCK_BOOK_STATE,
        selectedBook: "Book 2",
    });

    render(<Timeline />);

    fireEvent.click(screen.getByTestId("back-button"));

    expect(mockDispatch).toHaveBeenCalledWith(setSelectedBook("Book 1"));
});

test("handle back action - goes to previous chapter", () => {
    useSelector.mockReturnValue({
        ...MOCK_BOOK_STATE,
        selectedChapter: 1,
    });
    render(<Timeline />);

    fireEvent.click(screen.getByTestId("back-button"));

    expect(mockDispatch).toHaveBeenCalledWith(setSelectedChapter(0));
});

test("handle next action - stop at last book and chapter", () => {
    useSelector.mockReturnValue({
        ...MOCK_BOOK_STATE,
        selectedBook: "Book 2",
        selectedChapter: 1,
    });

    render(<Timeline />);

    fireEvent.click(screen.getByTestId("next-button"));

    expect(mockDispatch).not.toHaveBeenCalled();
});

test("handle next action - goes to next book", () => {
    useSelector.mockReturnValue({
        ...MOCK_BOOK_STATE,
        selectedChapter: 1,
    });

    render(<Timeline />);

    fireEvent.click(screen.getByTestId("next-button"));

    expect(mockDispatch).toHaveBeenCalledWith(setSelectedBook("Book 2"));
});

test("handle next action - goes to next chapter", () => {
    render(<Timeline />);

    fireEvent.click(screen.getByTestId("next-button"));

    expect(mockDispatch).toHaveBeenCalledWith(setSelectedChapter(1));
});

test("match snapshot", () => {
    const { container } = render(<Timeline />);

    expect(container).toMatchSnapshot();
});
