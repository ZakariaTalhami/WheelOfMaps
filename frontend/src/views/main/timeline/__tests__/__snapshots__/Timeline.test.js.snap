// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div>
  <div
    class="css-5uvpa0"
  >
    <div
      class="css-ug2oqc"
    >
      <button
        aria-label="Previous Chapter"
        class="chakra-button css-1a8ali6"
        data-testid="back-button"
        type="button"
      >
        <i
          aria-hidden="true"
          class="icon-back"
          focusable="false"
        />
      </button>
    </div>
    <div
      class="css-sjan41"
    >
      <div
        class="chakra-stack css-1mhc6c5"
      >
        <div
          class="chakra-select__wrapper css-1njiejo"
        >
          <select
            class="chakra-select css-wd7gg9"
            data-testid="book-select"
            options="Book 1,Book 2"
          >
            <option
              data-testid="option"
              style="background-color: rgb(169, 153, 133); appearance: none; padding: 10px;"
              value="Book 1"
            >
              Book 1
            </option>
            <option
              data-testid="option"
              style="background-color: rgb(169, 153, 133); appearance: none; padding: 10px;"
              value="Book 2"
            >
              Book 2
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-h22hlv"
          />
        </div>
        <div
          class="chakra-select__wrapper css-1njiejo"
        >
          <select
            class="chakra-select css-wd7gg9"
            data-testid="chapter-select"
            options="Chapter 1: title 1,Chapter 2: title 1"
          >
            <option
              data-testid="option"
              style="background-color: rgb(169, 153, 133); appearance: none; padding: 10px;"
              value="Chapter 1: title 1"
            >
              Chapter 1: title 1
            </option>
            <option
              data-testid="option"
              style="background-color: rgb(169, 153, 133); appearance: none; padding: 10px;"
              value="Chapter 2: title 1"
            >
              Chapter 2: title 1
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-h22hlv"
          />
        </div>
      </div>
    </div>
    <div
      class="css-ug2oqc"
    >
      <button
        aria-label="Next Chapter"
        class="chakra-button css-1a8ali6"
        data-testid="next-button"
        type="button"
      >
        <i
          aria-hidden="true"
          class="icon-next"
          focusable="false"
        />
      </button>
    </div>
  </div>
</div>
`;
